src/utils/AddressUtils.ts

```diff
@@ -68,3 +70,170 @@ export function isValidEvmAddress(address: string): boolean {
+// which for this class, is a bytes32 hex string.
+export class Address {
+  readonly rawAddress;
+  constructor(_rawAddress: Uint8Array) {
+    // The only validation done here is checking that the address is at most a 32-byte address, which  will be well-defined on any supported network.
+    // Further validation is done on the subclasses so that we can guarantee that, for example, an EvmAddress type will always contain a valid, 20-byte
+    // EVM address.
+    if (_rawAddress.length > 32) {
+      throw new Error(`Address ${utils.hexlify(_rawAddress)} cannot be longer than 32 bytes.`);
+    }
+    // Ensure all addresses in this class are internally stored as 32 bytes.
+    this.rawAddress = utils.zeroPad(_rawAddress, 32);
+  }
+
+  // Converts the address into a bytes32 string. Note that the output bytes will be lowercase  so that it matches ethers event data. This function will never
+  // throw since address length validation was done at construction time.
+  toBytes32(): string {
+    return utils.hexZeroPad(utils.hexlify(this.rawAddress), 32).toLowerCase();
+  }
+
+  // Converts the address (can be bytes32 or bytes20) to its base58 counterpart. This conversion will always succeed, even if the input address is not valid on Solana,
+  // as this address may be needed to represent an EVM address on Solana.
+  toBase58(): string {
+    return bs58.encode(this.rawAddress);
+  }
+
+  // Converts the address to a valid EVM address. If it is unable to convert the address to a valid EVM address for some reason, such as if this address
+  // is longer than 20 bytes, then this function will throw an error.
+  toAddress(): string {
+    const hexString = utils.hexlify(this.rawAddress);
+    const rawAddress = utils.hexZeroPad(utils.hexStripZeros(hexString), 20);
+    return utils.getAddress(rawAddress);
+  }
+
+  // Implements `Hexable` for `Address`. Needed for encoding purposes. This class is treated by default as a bytes32 primitive type, but can change for subclasses.
+  toHexString(): string {
+    return this.toBytes32();
+  }
+
+  // Checks if this address can be coerced into a bytes20 evm address. Returns true if it is possible and false otherwise.
+  isValidEvmAddress(): boolean {
+    try {
+      this.toAddress();
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  // Checks if this address can be coerced into a valid Solana address. Return true if possible and false otherwise.
+  isValidSvmAddress(): boolean {
+    return isSvmAddress(this.toBase58());
+  }
+
+  // Checks if the address is valid on the given chain ID.
+  isValidOn(chainId: number): boolean {
+    if (chainIsEvm(chainId)) {
+      return this.isValidEvmAddress();
+    }
+    return this.isValidSvmAddress();
+  }
+
+  // Checks if the object is an address by looking at whether it has an Address constructor.
+  static isAddress(obj: Record<string, unknown>): boolean {
```

Comment:
Does `obj` need to be a Record type, or is `unknown` sufficient?

```suggestion
  static isAddress(obj: unknown): boolean {
```

Reply:
Unknown is sufficient. I just wanted to be more verbose here https://github.com/across-protocol/sdk/pull/909/commits/92539941d0dcc6c8a974822b9c84d5c5e524a465